HDDS-6396. [Multi-Tenant] Merge and cleanup tenant group/role/policy tables, refactor protobuf messages and `isTenantAdmin`

smengcl opened the pull request:
## What changes were proposed in this pull request?

- Remove `TenantGroupTable` (Unused)
- Add new field `roleIds` in `OmDBAccessIdInfo` (`TenantAccessIdTable` value)
- Remove `TenantRoleTable`
- Rename protobuf message `OmDBAccessInfo` to `ExtendedAccessIdInfo`
- Use protobuf in `OmDBTenantInfo` and `OmDBKerberosPrincipalInfo`
- Rename `OmDBKerberosPrincipalInfo` to `OmDBUserPrincipalInfo`
- Move `isTenantAdmin()` from `OzoneManager` to `OMMultiTenantManagerImpl`
  - Fix a bug where `listUsersInTenant` wasn't checking admin permission properly (didn't count the `omAdminUsernames` wildcard case)
- Remove `TenantPolicyTable` (merged into `OmDBTenantInfo` as a new `policyNames` field)
- Rename `OmDBTenantInfo` to `OmDBTenantState`

Some table query changes maybe required for #3131 with this patch.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6396

## How was this patch tested?

- All existing test cases should pass.

smengcl commented:
CI passed on my fork: https://github.com/smengcl/hadoop-ozone/commits/HDDS-6396

Gotta spend some time to fix the flaky scm initialization issue in the upgrade test additon. Has failed a lot recently.

prashantpogde commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmDBAccessIdInfo.java:
```diff
@@ -42,16 +42,19 @@
     this.userPrincipal = userPrincipal;
     this.isAdmin = isAdmin;
     this.isDelegatedAdmin = isDelegatedAdmin;
+    this.roleId = roleId;
```
what if accessID has to be part of multiple roles ?

smengcl replied:
Thanks @prashantpogde . Very good point. I have addressed the issue by making the protobuf a `repeated` one.

errose28 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1414,14 +1414,41 @@ message TenantInfo {
+    optional bool isDelegatedAdmin = 4;
+}
+
+message TenantUserInfo {
```
I think we can pull this message's fields into the TenantGetUserInfoResponse.

smengcl replied:
Done

errose28 reviewed (commented):
Thanks for cleaning up this existing code @smengcl. We talked offline about removing the account namespace and policy table keys (since we will probably remove the policy table) as well, but I don't think those parts are present in the diff for an inline comment.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java:
```diff
@@ -378,9 +382,60 @@ public void deactivateUser(String accessID)
+   * Internal isTenantAdmin method that takes a username String instead of UGI.
+   */
+  private boolean isTenantAdmin(String username, String tenantId,
+      Boolean delegated) {
```
Good point. Not sure why I used wrapper object at the time. Done

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -3089,11 +3088,11 @@ public TenantUserList listUsersInTenant(String tenantId, String prefix)
-          && !omAdminUsernames.contains(userName)) {
-        throw new IOException("Only tenant and ozone admins can access this " +
-            "API. '" + userName + "' is not an admin.");
+      final UserGroupInformation ugi = ProtobufRpcEngine.Server.getRemoteUser();
```
`listUsersInTenant` is only accessible from CLI right now.

This is not exposed to S3 Gateway. In theory S3 Gateway could be extended to add this functionality and then yes, the `ugi` would become S3 Gateway's.

errose28 replied:
Oh that's right.

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantRequestHelper.java:
```diff
@@ -59,16 +59,17 @@ static void checkAdmin(OzoneManager ozoneManager) throws OMException {
   static void checkTenantAdmin(OzoneManager ozoneManager, String tenantId)
       throws OMException {
 
     final UserGroupInformation ugi = ProtobufRpcEngine.Server.getRemoteUser();
```
See above, not a problem.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java:
```diff
@@ -378,9 +382,60 @@ public void deactivateUser(String accessID)
+      for (final String accessId : principalInfo.getAccessIds()) {
+        final OmDBAccessIdInfo accessIdInfo =
+            omMetadataManager.getTenantAccessIdTable().get(accessId);
+        if (tenantId.equals(accessIdInfo.getTenantId())) {
```
Ok. Added null check

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/OMDBDefinition.java:
```diff
@@ -199,15 +199,14 @@
+                    new OmDBUserPrincipalInfoCodec());
 
   public static final DBColumnFamilyDefinition<String, OmDBTenantInfo>
             TENANT_STATE_TABLE =
```
Done

smengcl commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:
```diff
@@ -345,7 +345,11 @@ private OzoneConsts() {
+  // Tenant user role. All tenant users are assigned this role by default
+  public static final String TENANT_ROLE_USER = "user";
+  // Tenant admin role. All tenant admins are assigned this role
+  public static final String TENANT_ROLE_ADMIN = "admin";
```
Done

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmDBAccessIdInfo.java:
```diff
@@ -42,16 +47,23 @@
+   * e.g. OzoneConsts.TENANT_ROLE_USER, OzoneConsts.TENANT_ROLE_ADMIN,
+   *      or other custom role names.
+   */
+  private final Set<String> roleIds;
```
I use "Id" suffix here just to be consistent with others:

```
tenantId = tenant name
policyId = policy name
roleId   = role name
```

Yeah it is a bit misleading as none of those "Id"s means "number".. 

Maybe change all of them (back) to a "Name" suffix?

smengcl replied:
For PolicyID, we do use the number, but stored as String: 

https://github.com/apache/ozone/blob/8ecff6fd1a5842e8c81c125fd1d5aa729b7c7a47/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerAccessPolicy.java#L72

If we change it to `long` we should change the type in `OmDBTenantState` as well.

errose28 replied:
Seems odd that we use name for role and ID for policy. I think we should standardize and use either IDs for role and policy, or names for role and policy. When a policy is created, roles can be created in the same operation, however you will only get the policy ID back. This means we need to track roles by name, so I think we should track policies by name as well. I think this will make error handling, retries, and tests for existence easier as well.

I am currently updating the `MultiTenantAccessController` to operate on names instead of IDs.

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmDBTenantInfo.java:
```diff
@@ -152,6 +109,33 @@ public String getBucketPolicyGroupName() {
+    return OzoneManagerProtocolProtos.TenantInfo.newBuilder()
+        .setTenantId(tenantId)
+        .setBucketNamespaceName(bucketNamespaceName)
+        .setAccountNamespaceName(accountNamespaceName)
```
Removed

errose28 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1409,19 +1409,39 @@ message SetS3SecretResponse {
-    optional string accountNamespaceName = 3;
-    optional string userPolicyGroupName = 4;
-    optional string bucketPolicyGroupName = 5;
+    repeated string policyNames = 3;
```
Should we track the roles associated with each tenant here?

smengcl replied:
We could. Should we do that?
We didn't have tenant -> roleNames mapping before.

Unless:
1. we are intending to store some roleIndex -> roleName mapping in TenantInfo (TenantState) and use roleIndex in ExtendedAccessIdInfo to reduce redundancy. e.g. tenant1-UserRole = index 0, tenant1-AdminRole = index 1.
2. it is required to iterate all role names assigned in this tenant quickly
3. the Ranger sync thread would need it when deleting a tenant

(3) might be valid? @prashantpogde

errose28 reviewed (commented):
Thanks for the updates @smengcl, a few more comments inline.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManager.java:
```diff
@@ -216,6 +221,24 @@ TenantUserList listUsersInTenant(String tenantID, String prefix)
+   * @param tenantId tenant name
+   * @return user role name. e.g. tenant1-UserRole
+   */
+  static String getUserRoleId(String tenantId) {
```
Yup. Missed this one. Thanks for pointing that out.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantRequestHelper.java:
```diff
@@ -182,8 +184,10 @@ public static boolean isUserAccessIdPrincipalOrTenantAdmin(
+    // Check if ugi is a tenant admin (or an Ozone cluster admin)
+    final OMMultiTenantManager multiTenantManager =
+        ozoneManager.getMultiTenantManager();
+    if (multiTenantManager.isTenantAdmin(ugi, tenantId, true)) {
```
Good point. Should be false.

btw This helper is only used in SetSecret right now.

smengcl replied:
When set to false, a non-delegated admin could also SetSecret of a delegated admin.

This should not pose security issue, as the AccessId/Key pair is only used to access via S3 Gateway, but not any other administrative tasks that requires CLI access (that requires separate Kerberos auth).

smengcl commented:
> Thanks for the updates @smengcl, a few more comments inline.

Thanks for the 2nd round review @errose28 . I have addressed most comments.

Pending discussion about the `roleNames` field in `TenantState`.

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java:
```diff
@@ -292,27 +291,27 @@ public OMClientResponse validateAndUpdateCache(
-
-      final OmDBTenantState omDBTenantInfo = new OmDBTenantState(
-          tenantId, bucketNamespaceName, policyNamesList);
+          OMMultiTenantManager.getDefaultBucketNamespacePolicyName(tenantId);
```
Should be `OMMultiTenantManager.getDefaultBucketPolicyName(tenantId);`?

smengcl replied:
Good catch

errose28 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOMMultiTenantManagerImpl.java:
```diff
@@ -65,18 +62,22 @@ public void setUp() throws IOException {
+    final String bucketNamespacePolicyName =
+        OMMultiTenantManager.getDefaultBucketNamespacePolicyName(tenantName);
+    final String bucketPolicyName =
+        OMMultiTenantManager.getDefaultBucketNamespacePolicyName(tenantName);
```
Same as above, this looks like it should be `OMMultiTenantManager.getDefaultBucketPolicyName(tenantName);`

errose28 reviewed (commented):
Just two minor things @smengcl otherwise LGTM. Thanks for the offline discussions on proto changes.

smengcl commented:
Thanks @errose28 for the discussion and review. I have addressed the last issue. :D

errose28 reviewed (approved):
Thanks for the last quick fix @smengcl LGTM.

smengcl commented:
Thanks again @errose28 for the review and +1. Will merge shortly.
